compiler/rustc_middle/src/ty/consts/int.rs

```diff
@@ -116,41 +115,46 @@ impl std::fmt::Debug for ConstInt {
 
 // FIXME: reuse in `super::int::ConstInt` and `Scalar::Bits`
 /// The raw bytes of a simple value.
+///
+/// This is a packed struct in order to allow this type to be optimally embedded in enums
+/// (like Scalar).
 #[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash)]
-#[derive(HashStable)]
+#[repr(packed)]
 pub struct ScalarInt {
     /// The first `size` bytes of `data` are the value.
     /// Do not try to read less or more bytes than that. The remaining bytes must be 0.
-    ///
-    /// This is an array in order to allow this type to be optimally embedded in enums
-    /// (like Scalar).
-    bytes: [u8; 16],
+    data: u128,
     size: u8,
 }
 
+// Cannot derive these, as the derives take references to the fields, and we
+// can't take references to fields of packed structs.
+impl<CTX> crate::ty::HashStable<CTX> for ScalarInt {
+    fn hash_stable(&self, hcx: &mut CTX, hasher: &mut crate::ty::StableHasher) {
+        {self.data}.hash_stable(hcx, hasher);
```

Comment:
Maybe add a comment that the `{}` causes a temporary to be borrowed instead of `self.data` directly?

Reply:
I'd say something like "force copy instead of borrow".

Reply:
I left some elaborate explanations at each site